src/nunavut/lang/c/templates/_serialize.j2

```diff
@@ -1,24 +1,108 @@
-{%- macro canard_set(field_name) -%}
-out_buffer, offset + length, in_instance->{{ field_name | id }}[i]
+{#
+    Handy macro to populate common first three fields in base serilization functions.
+#}
+{%- macro nunavut_set(field_name, ref) -%}
+out_buffer, offset + length, {{ ref }}{{ field_name | id }}[i]
 {%- endmacro %}
 
-{%- macro set_basic(loop_type, field, include_bool) -%}
-    {%- if loop_type is UnsignedIntegerType -%}
-    nunavutSetUxx({{ canard_set(field.name) }}, {{ loop_type.bit_length }});
-    {%- elif loop_type is SignedIntegerType -%}
-    nunavutSetIxx({{ canard_set(field.name) }}, {{ loop_type.bit_length }});
-    {%- elif loop_type is FloatType -%}
-    nunavutSetF{{ loop_type.bit_length }}({{ canard_set(field.name) }});
-    {%- endif -%}
-    {%- if include_bool and loop_type is BooleanType -%}
-    nunavutSetBit({% include '_canard_set.j2' %});
-    {%- endif -%}
+{#-
+    Entry macro, call this from main templates
+    Differs from _serialize_composite because we need to handle top-level type case
+#}
+{%- macro serialize(type) %}
+    {%- if type is StructureType %}
+    {%- for f, offset in type.iterate_fields_with_offsets(0) %}
+    // Begin Structure: {{ f.data_type }}
+    {{ _serialize_any(f, 'in_instance->', offset) }}
+    // End Structure: {{ f.data_type }}
+    {%- endfor %}
+    {%- elif type is UnionType %}
+    // Begin Union: {{ type }}
+    {%- for f in type.fields %}
+    {%- if loop.first %}
+    if ({{ type | full_reference_name }}__is_{{ f.name }}(in_instance)))
+    {%- else %}
+    else if ({{ type | full_reference_name }}__is_{{ f.name }}(in_instance)
+    {%- endif %}
+    {
+        {{ _serialize_any(f, 'in_instance->', offset) }}
+    }
+    {%- endfor %}
+    else
+    {
+        return NUNAVUT_ERR_INVALID_TAG;
+    }
+    // End Union: {{ type }}
+    {%- endif %}
 {%- endmacro %}
 
-{%- macro field(field) %}
+{%- macro _serialize_any(field, ref, offset) %}
     {#- TODO formatting is slightly off somewhere here,
         includes an extra line before the content #}
-    {%- if field.data_type is ArrayType %}
+    {%- if field.data_type is PrimitiveType %}
+    {{ _serialize_primitive(field, ref) }}
+    {%- elif field.data_type is ArrayType %}
+    {{ _serialize_array(field, ref, offset) }}
+    {%- elif field.data_type is CompositeType %}
+    {#- Jinja yells at me here that offset is 'undefined'... no idea why #}
```

Comment:
This one is odd. It will correctly recognize that offset was not provided if I don't add it in the macro call, but I get an exception here.

I've tried nothing and I'm all out of ideas.

Reply:
¯\\\_(ツ)\_/¯

Reply:
Resolution: offset was undefined in a higher-level call to `_serialize_any`. (specifically line 21 in this commit)

Why it wasn't caught there... ¯\_(ツ)_/¯